context/src/bazel_bep/parser.rs

```diff
@@ -0,0 +1,155 @@
+use anyhow::Ok;
+use bazel_bep::types::build_event_stream::{build_event::Payload, file::File::Uri, BuildEvent};
+use serde_json::Deserializer;
+
+#[derive(Debug, Clone)]
+pub struct TestResult {
+    pub cached: bool,
+    pub xml_files: Vec<String>,
+}
+
+const FILE_URI_PREFIX: &str = "file://";
+
+/// Uses proto spec
+/// https://github.com/TylerJang27/bazel-bep/blob/master/proto/build_event_stream.proto based on
+/// https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/buildeventstream/proto/build_event_stream.proto
+pub struct BazelBepParser {
+    bazel_bep_path: String,
+    test_results: Vec<TestResult>,
+    errors: Vec<String>,
+}
+
+impl BazelBepParser {
+    pub fn new(bazel_bep_path: String) -> Self {
+        Self {
+            bazel_bep_path,
+            test_results: Vec::new(),
+            errors: Vec::new(),
+        }
+    }
+
+    pub fn errors(&self) -> &Vec<String> {
+        &self.errors
+    }
+
+    pub fn xml_files(&self) -> Vec<String> {
+        self.test_results
+            .iter()
+            .filter_map(|r| {
+                if r.cached {
+                    return None;
+                }
+                Some(r.xml_files.clone())
+            })
+            .flatten()
+            .collect()
+    }
+
+    pub fn parse(&mut self) -> anyhow::Result<()> {
+        let file = std::fs::File::open(&self.bazel_bep_path)?;
+        let reader = std::io::BufReader::new(file);
+
+        let build_events = Deserializer::from_reader(reader).into_iter::<BuildEvent>();
+        build_events.for_each(|parse_event| {
```

Comment:
`for_each` is for side-effects, I'd recommend using a normal `for` loop or a `fold` here.

For example:

```rust
let (errors, test_results) = Deserializer::from_reader(reader)
    .into_iter::<BuildEvent>()
    .fold(
        (Vec::<String>::new(), Vec::<TestResult>::new()),
        |(mut errors, mut test_results), parse_event| {
            match parse_event {
                Result::Err(ref err) => {
                    errors.push(format!("Error parsing build event: {}", err));
                }
                Result::Ok(build_event) => {
                    if let Some(Payload::TestResult(test_result)) = build_event.payload {
                        let xml_files = test_result
                            .test_action_output
                            .into_iter()
                            .filter_map(|action_output| {
                                if action_output.name.ends_with(".xml") {
                                    action_output.file.and_then(|f| {
                                        if let Uri(uri) = f {
                                            Some(
                                                uri.strip_prefix(FILE_URI_PREFIX)
                                                    .unwrap_or(&uri)
                                                    .to_string(),
                                            )
                                        } else {
                                            None
                                        }
                                    })
                                } else {
                                    None
                                }
                            })
                            .collect();

                        let cached =
                            if let Some(execution_info) = test_result.execution_info {
                                execution_info.cached_remotely || test_result.cached_locally
                            } else {
                                test_result.cached_locally
                            };

                        test_results.push(TestResult { cached, xml_files });
                    }
                }
            }

            (errors, test_results)
        },
    );

self.errors = errors;
self.test_results = test_results;
```